### PR TITLE
winch: Refactor the prologue and epilogue interface

### PR DESCRIPTION
**Pull request by elliottt**
Rework the winch prologue and epilogue definitions as default implementations in the `MacroAssembler` trait, naming the common sequence of frame setup, check stack, handle clobbers, and reserve space for locals. This change also allows us to remove the special prologue and epilogue functions from the trampoline module, and rely instead on the same function used by the normal function compilation path.

As I moved the pinning of the vmctx register to after the prologue, this change had a pretty noisy effect on the filetests.

### CODE REVIEW

File: winch/codegen/src/trampoline.rs [hunk 1]
```diff
@@ -230,7 +231,9 @@ where
         if wasm_sig.has_stack_results() {
             self.masm.free_stack(wasm_sig.results.size());
         }
-        self.epilogue_with_callee_saved_restore(spill_size);
+
+        self.masm.free_stack(spill_size);
+        self.masm.epilogue(&self.callee_saved_regs, 0);
```
Comment (saulecabrera) [inline winch/codegen/src/trampoline.rs:236]: Could we do `self.masm.epilogue(&self.callee_saved_regs, spill_size)` here?

**Review by saulecabrera — APPROVED** [review]
One minor question. Else looks reasonable to me, thanks!

File: winch/codegen/src/trampoline.rs (hunk 1 above)
Comment (elliottt) [reply winch/codegen/src/trampoline.rs:236]: We could, but at that point i wonder if we should move the stack allocation out of  the prologue and epilogue. What do you think about not including the reserve/free calls?

Comment (saulecabrera) [reply winch/codegen/src/trampoline.rs]: Yeah, that makes sense to me as well.

**Pull request merged**